plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -60,120 +64,110 @@ becomes
 let f = \x.\y -> x
 in
   let z = f q
-  in ((\x.\y -> x) a b)
-
-With beta reduction, it becomes:
-
-let f = \x.\y -> x
-in
-  let z = f q
-  in a (more accurately it becomes (let { x = a, y = b } in x))
-
-This is already a reduction in code size. However, because of this,
-our dead code elimination pass is able to further reduce the code to just `a`.
-
-Because a function can be called in the `body` multiple times and may not be fully applied for all
-its calls, we cannot simply keep a substitution map like in `Inline`,
-which substitute *all* occurrences of a variable. Instead, we store all in scope variables in a
-map, `Utils.NonRecInScopeSet`, with all information needed for inlining saturated functions.
-
-To find out whether a function is fully applied, when we encounter a variable in the `body`,
-we find its arity from the `Utils.NonRecInScopeSet` map and compare it with the list
-of arguments it has been applied to at that site. If it is fully applied, we inline it there.
-
-Note that over-application of a function would also be inlined. E.g.:
-
-```haskell
-let id = \y -> y
-     f = \x -> id
- in f x y
-```
+  in a
 
-`f`'s arity is (\x) or 1, but is applied to 2 arguments. We inline `f` in this case if its cost and
-size are acceptable.
+Note that because `f` occurs twice in the original term, and the RHS of `f` is not small enough,
+`f` is not unconditionally inlined. Instead, the second usage of `f` is inlined. The first
+usage is not inlined because it is not fully applied.
 
-(2) How do we decide whether cost and size are acceptable?
+(2) How do we decide whether a fully saturated application can be inlined?
 
-We currently reuse the heuristics 'Utils.sizeIsAcceptable' and 'Utils.costIsAcceptable'
-that are used in unconditional inlining. For
+For size, we compare the sizes (in terms of AST nodes before and after the inlining), and
+inline only if it does not increase the size. In the above example, we count the number of AST
+nodes in `f a b` and in `a`. The latter is smaller, which means inlining reduces the size.
 
-let v = \x1.\x2...\xn.VBody in body
+For cost, we check whether the RHS (in this example, `\x. \y -> x`) has a small cost.
+If the RHS has a non-zero arity, then the cost is always small (since a lambda or a type
```

Comment:
I think we can just mention [costIsAcceptable](https://github.com/input-output-hk/plutus/blob/2d0a953ffcdce0975595c3c9422ff1202d7ae63e/plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Utils.hs#L328) here. I don't see why we should discuss the case of lambda separately.

```suggestion
For cost, we check whether the RHS (in this example, `\x. \y -> x`) has an acceptable cost as specified in `Utils.costIsAcceptable`. 
```
```

Reply:
I remember writing something similar and both of you didn't think it was necessary! I guess I was right after all!